libensemble/generators.py

```diff
@@ -0,0 +1,206 @@
+                return create_points(num_points, self.param)
+
+            def tell(self, results):
+                self.model = update_model(results, self.model)
+
+            def final_tell(self, results):
+                self.tell(results)
+                return list(self.model)
+
+
+        my_generator = MyGenerator(my_parameter=100)
+        gen_specs = GenSpecs(generator=my_generator, ...)
+    """
+
+    @abstractmethod
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the Generator object on the user-side. Constants, class-attributes,
+        and preparation goes here.
+
+        .. code-block:: python
+
+            my_generator = MyGenerator(my_parameter, batch_size=10)
+        """
+
+    @abstractmethod
+    def ask(self, num_points: Optional[int]) -> List[dict]:
+        """
+        Request the next set of points to evaluate.
+        """
+
+    def ask_updates(self) -> List[npt.NDArray]:
+        """
+        Request any updates to previous points, e.g. minima discovered, points to cancel.
+        """
+
+    def tell(self, results: List[dict]) -> None:
+        """
+        Send the results of evaluations to the generator.
+        """
+
+    def final_tell(self, results: List[dict], *args, **kwargs) -> Optional[npt.NDArray]:
+        """
+        Send the last set of results to the generator, instruct it to cleanup, and
+        optionally retrieve an updated final state of evaluations. This is a separate
+        method to simplify the common pattern of noting internally if a
+        specific tell is the last. This will be called only once.
+        """
+
+
+class LibensembleGenerator(Generator):
+    """Internal implementation of Generator interface for use with libEnsemble, or for those who
+    prefer numpy arrays. ``ask/tell`` methods communicate lists of dictionaries, like the standard.
+    ``ask_numpy/tell_numpy`` methods communicate numpy arrays containing the same data.
+    """
+
+    def __init__(
+        self, History: npt.NDArray = [], persis_info: dict = {}, gen_specs: dict = {}, libE_info: dict = {}, **kwargs
+    ):
+        self.gen_specs = gen_specs
+        if len(kwargs) > 0:  # so user can specify gen-specific parameters as kwargs to constructor
+            self.gen_specs["user"] = kwargs
+        if not persis_info:
+            self.persis_info = add_unique_random_streams({}, 4, seed=4321)[1]
+        else:
+            self.persis_info = persis_info
+
+    @abstractmethod
+    def ask_numpy(self, num_points: Optional[int] = 0) -> npt.NDArray:
+        """Request the next set of points to evaluate, as a NumPy array."""
+
+    @abstractmethod
+    def tell_numpy(self, results: npt.NDArray) -> None:
+        """Send the results, as a NumPy array, of evaluations to the generator."""
+
+    def ask(self, num_points: Optional[int] = 0) -> List[dict]:
+        """Request the next set of points to evaluate."""
+        return np_to_list_dicts(self.ask_numpy(num_points))
```

Comment:
I think we will want to convert from numpy types to regular python types. This could be done inside np_to_list_dicts or here e.g.

```python
    @staticmethod
    def convert_np_types(dict_list):
        np_types = (np.float64, np.int64, np.int32, np.float32)
        return [
            {key: (value.item() if isinstance(value, np_types) else value) for key, value in item.items()}
            for item in dict_list
        ]

    def ask(self, num_points: Optional[int] = 0) -> List[dict]:
        """Request the next set of points to evaluate."""
        vals = LibensembleGenerator.convert_np_types(np_to_list_dicts(self.ask_numpy(num_points)))
        return vals
```